### PR TITLE
Added good flow unit tests

### PR DESCRIPTION
**Pull request by shlomimatichin**

### CODE REVIEW

File: prettier.config.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+module.exports = {
```
Comment (roni-frantchi) [inline prettier.config.js]: Please rename to `.prettierrc` and use this config:  
```
{
  "semi": true,
  "singleQuote": true,
  "printWidth": 120,
  "trailingComma": "none",
  "parser": "typescript",
  "proseWrap": "preserve",
  "quoteProps": "as-needed",
  "doc": false,
  "jsxBracketSameLine": true,
  "arrowParens": "avoid"
}
```

File: lib/terratag-action.spec.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+const mockedChildProcess = childProcess as jest.Mocked<typeof childProcess>;
+import core from "@actions/core";
+const mockedCore = core as jest.Mocked<typeof core>;
+import tc from "@actions/tool-cache";
+const mockedTC = tc as jest.Mocked<typeof tc>;
+
+import run from "./terratag-action";
+
+test("simple end to end, latest terratag", async () => {
+  jest.resetAllMocks();
```
Comment (roni-frantchi) [inline lib/terratag-action.spec.ts]: Please configure jest to reset mocks for you instead of doing that on each test.  
https://jestjs.io/docs/en/configuration#resetmocks-boolean

File: lib/terratag-action.spec.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+test("simple end to end, specific terratag", async () => {
+  jest.resetAllMocks();
+  mockedCore.getInput.mockImplementation((name: string) => {
+    const inputs: { [key: string]: string } = { terratagVersion: "5.6.7", tags: JSON.stringify({ a: "b" }) };
+    return inputs[name];
+  });
+  mockedAxios.get.mockRejectedValue("Should not be called");
+  mockedTC.downloadTool.mockResolvedValue("FAKE PATH FOR DOWNLOADED TOOL TAR");
+  mockedTC.extractTar.mockResolvedValue("FAKE PATH FOR EXTRACTED");
```
Comment (roni-frantchi) [inline lib/terratag-action.spec.ts]: All these are copy-pasted on some of these tests.  
Please use a proper `describe()` hierarchy to group those and use `beforeEach()` to mock on each test with common mocks.

File: lib/terratag-action.spec.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
+jest.mock("axios");
+jest.mock("child_process");
```
Comment (roni-frantchi) [inline lib/terratag-action.spec.ts]: jest hoists these for you, let's keep `import`s on top

File: lib/terratag-action.spec.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+import axios from "axios";
+const mockedAxios = axios as jest.Mocked<typeof axios>;
+import childProcess from "child_process";
+const mockedChildProcess = childProcess as jest.Mocked<typeof childProcess>;
+import core from "@actions/core";
+const mockedCore = core as jest.Mocked<typeof core>;
+import tc from "@actions/tool-cache";
+const mockedTC = tc as jest.Mocked<typeof tc>;
```
Comment (roni-frantchi) [inline lib/terratag-action.spec.ts]: Let's keep `import`s on top

File: lib/terratag-action.spec.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  });
+  mockedChildProcess.spawn.mockReturnValue(spawn as any);
+  await run();
+  expect(mockedCore.addPath.mock.calls).toEqual([["FAKE PATH FOR EXTRACTED"]]);
+  expect(mockedAxios.get.mock.calls).toEqual([["https://github.com/env0/terratag/releases"]]);
+  expect(mockedTC.downloadTool.mock.calls).toEqual([
+    ["https://github.com/env0/terratag/releases/download/v1.2.3/terratag_1.2.3_linux_amd64.tar.gz"],
+  ]);
+  expect(mockedTC.extractTar.mock.calls).toEqual([["FAKE PATH FOR DOWNLOADED TOOL TAR"]]);
+  expect(mockedChildProcess.spawn.mock.calls).toEqual([["FAKE PATH FOR EXTRACTED/terratag", ['-tags={"a":"b"}']]]);
```
Comment (roni-frantchi) [inline lib/terratag-action.spec.ts]: 1. Let's use the more commonly accepted bdd jargon of `it('should do this')`
2. It's better to have one `expect` per `it` - so that if one fail the rest continue to run - you should separate the the "given"/"when" part from the "then".  

Something like:
```typescript
describe('terratag action', () => {
  describe('on success', () => {
    beforeEach(() => { /* mock a happy path and run() */ })
    describe('downloading terratag', () => {
          it('should download the latest terratag version', () => {/* whatever you expect */});
          it('should extract the downloaded tar', () => {/* whatever you expect */});
    });

    it('should spawn a process invoking terratag', () => {});

    describe('supported flags', () => {
      it('should set the dir flag', () => {});
      /* ... */
    });
  });
});
```

Same for errors.

**Review by roni-frantchi — COMMENTED** [review]
Well done adding ci and prettier.  
Let's revise these tests to comply with best practices

File: lib/terratag-action.spec.ts (hunk 6 above)
Comment (shlomimatichin) [reply lib/terratag-action.spec.ts]: lost you here. sorry. must be missing something:
1. if mocks are reset between every it, how can i expect in a different `it` than where it run?
2. if i have different mock setup between cases, how can the `run` be executed in the `beforeEach`?

Comment (roni-frantchi) [reply lib/terratag-action.spec.ts]: You use `beforeEach` on different levels to set different mocks on different `describe` levels.  
It's alright to call `run()` several times after having mocked it differently of course, but having each `expect()` in a separate `it()`, you'll have shared mocks and run calls for sure.

File: __mocks__/@actions/core.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+const self = {
+    error: jest.fn(),
+    info: jest.fn(),
+    debug: jest.fn(),
+    warn: jest.fn(),
+    getInput: jest.fn(),
+    addPath: jest.fn(),
+}
+export default self;
```
Comment (roni-frantchi) [inline __mocks__/@actions/core.ts:9]: Why did you need to specify those? 
Didn't jest automocked these for you?

**Review by roni-frantchi — APPROVED** [review]

File: __mocks__/@actions/core.ts (hunk 7 above)
Comment (shlomimatichin) [reply __mocks__/@actions/core.ts:9]: nope. @actions/core does some type voodoo, mocking it with jest without this file causes errors such as "undefined does not have `addPath` property". i didn't research why that is.
